src/Viewer/services/decoder/FileManager.js

```diff
@@ -176,6 +190,38 @@ class FileManager {
         this.state.numberOfEvents = this._logEventOffsets.length;
     };
 
+    /**
+     * Given a ref timestamp, find the first log event whose timestamp is 
+     * greater or equal to the ref timestamp.
+     * Since all the log events are naturally sorted w.r.t. the timestamp,
+     * we can use binary search.
+     * Return the index of the log event.
+     * If the log event doesn't exist, the function should return the total
+     * number of events.
+     */
+    _findFirstLogEventWithTimestamp (refTimestamp) {
+        const numberOfEvents = this._logEventOffsets.length;
+        let left = 0;
+        let right = numberOfEvents - 1;
+        let mid;
+
+        while (left <= right) {
+            mid = left + Math.floor((right - left) / 2); // To avoid overflow.
+            if (this._logEventOffsets[mid].timestamp >= refTimestamp) {
+                if (mid === 0 || this._logEventOffsets[mid - 1].timestamp < refTimestamp) {
+                    return mid;
+                } else {
+                    right = mid - 1;
+                }
+            } else {
+                left = mid + 1;
+            }
+        }
+
+        // Not found
+        return numberOfEvents;
```

Comment:
Since we are returning the index of the log event that was found, if no event was found, do you think we could return the index of the last event rather than the number of events to keep it consistent? Then in decompressAndLoadFile we don't have to have the extra if statement, the new logEventIdx will be the returned index plus 1.

Reply:
Alternatively, we can return mid+1, so the function will return the log event number in both cases

Reply:
The function might be reused in other cases. Returning the actual array index might be more general. However, when nothing's found, it should return null instead of numberOfEvents.